rqt_py_common/CMakeLists.txt

```diff
@@ -25,14 +25,30 @@ if(BUILD_TESTING)
     set(${PROJECT_NAME}_PYTHON_EXECUTABLE "${PYTHON_EXECUTABLE}")
   endif()
 
+  # use different project name for generated test interfaces in order to make
+  # them importable in the tests beside the Python module of this package
+  set(_PROJECT_NAME ${PROJECT_NAME})
+  set(PROJECT_NAME "rqt_py_common_test_interfaces")
   rosidl_generate_interfaces(${PROJECT_NAME}
     test/msg/Val.msg
     test/msg/ArrayVal.msg
     test/srv/AddTwoInts.srv
-    DEPENDENCIES)
+    SKIP_INSTALL)
+  # setup fake ament index
+  file(MAKE_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/ament_index/share/ament_index/resource_index/packages")
+  file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/ament_index/share/ament_index/resource_index/packages/${_PROJECT_NAME}")
+  file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/ament_index/share/ament_index/resource_index/packages/${PROJECT_NAME}")
+  file(MAKE_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/ament_index/share/ament_index/resource_index/rosidl_interfaces")
+  file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/ament_index/share/ament_index/resource_index/rosidl_interfaces/${_PROJECT_NAME}")
+  file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/ament_index/share/ament_index/resource_index/rosidl_interfaces/${PROJECT_NAME}"
+    "msg/Val.msg\n"
+    "msg/ArrayVal.msg\n"
+    "srv/AddTwoInts.srv")
+  set(PROJECT_NAME ${_PROJECT_NAME})
 
   ament_add_pytest_test(${PROJECT_NAME} test
-    APPEND_ENV PYTHONPATH=${CMAKE_CURRENT_BINARY_DIR}
+    ENV "AMENT_PREFIX_PATH=${CMAKE_CURRENT_BINARY_DIR}/ament_index"
```

Comment:
Wouldn't using `APPEND_ENV` allow you to avoid creating fake markers for the real project?

Reply:
Unfortunately no.

First the tests need the markers for the fake project `rqt_py_common_test_interfaces`.

And second the code is written so poorly that it would fail if package `rqt_py_common` is in the index but doesn't install interfaces. Therefore appending the custom prefix path isn't sufficient since `rqt_py_common` would be found before. Atm there is no `PREPEND_ENV` option which would at least avoid the second problem.

Reply:
Ah, unfortunate.

In that case, is this an option?
```
ENV "AMENT_PREFIX_PATH=${CMAKE_CURRENT_BINARY_DIR}/ament_index;$ENV{AMENT_PREFIX_PATH}"
```

Reply:
Yes, I just results in a very large command for the test invocation. That is why I scaled it back to only what is needed since the tests don't mind about the other prefix path.

Reply:
Thanks for explaining.

Reply:
Np, the fact that we need this kind of workaround is already unfortunate...